### PR TITLE
verification of pkg/addr

### PR DESCRIPTION
**Pull request by aaronbojarski**
Verifies the pkg/addr package. 

The package was actually simplified in the new `scionproto/scion` code. It uses fewer interfaces. However the package now uses the net/netip go standard library.

Also `serrors.WrapStr` no longer exists. Therefore the gobra annotations will also be removed in this PR.

### CODE REVIEW

File: verification/dependencies/net/netip/netip.gobra [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  	def "github.com/scionproto/scion/verification/utils/definitions"
+)
+
+
+type Addr struct {
+	privateField def.PrivateField
+}
+
+decreases
+func (ip Addr)String() string
```
Comment (jcp19) [inline verification/dependencies/net/netip/netip.gobra]:
```suggestion
decreases _
func (ip Addr) String() string
```

Comment (jcp19) [reply verification/dependencies/net/netip/netip.gobra]: Termination is assumed, not proven, so we should use `decreases _`

File: verification/dependencies/net/netip/netip.gobra [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+ensures err != nil ==> err.ErrorMem()
+decreases
```
Comment (jcp19) [inline verification/dependencies/net/netip/netip.gobra]:
```suggestion
decreases _
```

File: verification/dependencies/net/net.gobra [hunk 3]
```diff
@@ -152,3 +152,7 @@ type Listener interface {
 	decreases
 	Addr() (a Addr)
 }
+
+ensures err != nil ==> err.ErrorMem()
+decreases
```
Comment (jcp19) [inline verification/dependencies/net/net.gobra]:
```suggestion
decreases _
```

File: pkg/addr/svc.go [hunk 4]
```diff
@@ -46,6 +48,7 @@ type SVC uint16
 // SVC addresses, use CS_A and DS_A; shorthand versions without
 // the _A suffix (e.g., CS) also return anycast SVC addresses. For multicast,
 // use CS_M, and DS_M.
+// @ decreases
 func ParseSVC(str string) (SVC, error) {
```
Comment (jcp19) [inline pkg/addr/svc.go]: this should return error.Mem() if the error is not nil

File: pkg/addr/isdas.go [hunk 5]
```diff
@@ -232,6 +245,8 @@ func ParseIA(ia string) (IA, error) {
 
 // MustParseIA parses s and returns the corresponding addr.IA object. It
 // panics if s is not a valid ISD-AS representation.
+// @ requires false
+// @ decreases
```
Comment (jcp19) [inline pkg/addr/isdas.go]:
```suggestion
```

We don't care for termination of functions that we can't call

File: pkg/addr/isdas.go [hunk 6]
```diff
@@ -213,8 +225,9 @@ func IAFrom(isd ISD, _as AS) (ia IA, err error) {
 }
 
 // ParseIA parses an IA from a string of the format 'isd-as'.
+// @ ensures err != nil ==> err.ErrorMem()
 // @ decreases
-func ParseIA(ia string) (IA, error) {
+func ParseIA(ia string) (ret_ia IA, err error) {
```
Comment (jcp19) [inline pkg/addr/isdas.go]: underscores separating parts of var ids is very unidiomatic in Go/Gobra. Please use camel case

File: pkg/addr/isdas.go [hunk 7]
```diff
@@ -80,22 +85,28 @@ type AS uint64
...
+// @ ensures retErr != nil ==> retErr.ErrorMem()
 // @ decreases
 func ParseAS(_as string) (retAs AS, retErr error) {
 	return parseAS(_as, ":")
 }
 
 // MustParseAS parses s and returns the corresponding addr.AS object. It panics
 // if s is not valid AS representation.
+// @ requires false
+// @ decreases
```
Comment (jcp19) [inline pkg/addr/isdas.go]:
```suggestion
```

File: pkg/addr/isdas.go [hunk 8]
```diff
@@ -67,6 +70,8 @@ func MustParseISD(s string) ISD {
 	return isd
 }
 
+// @ requires isd >= 0
```
Comment (jcp19) [inline pkg/addr/isdas.go]: given that an `isd` is a `uint16`, we know this property holds all the time (even though Gobra cannot prove it right now). As such, I would not put this in the specification, as we plan to fix this in Gobra. If you really need this fact, you should introduce an assumption instead

File: pkg/addr/isdas.go [hunk 9]
```diff
@@ -59,6 +60,8 @@ func ParseISD(s string) (ISD, error) {
 
 // MustParseISD parses s and returns the corresponding addr.ISD object. It panics
 // if s is not valid ISD representation.
+// @ requires false
+// @ decreases
```
Comment (jcp19) [inline pkg/addr/isdas.go]:
```suggestion
```

File: pkg/addr/isdas.go [hunk 10]
```diff
@@ -48,8 +48,9 @@ type ISD uint16
 
 // ParseISD parses an ISD from a decimal string. Note that ISD 0 is parsed
 // without any errors.
+// @ ensures err != nil ==> err.ErrorMem()
 // @ decreases
-func ParseISD(s string) (ISD, error) {
+func ParseISD(s string) (ret_isd ISD, err error) {
```
Comment (jcp19) [inline pkg/addr/isdas.go]: same comment about the usage of `_`, see my comment below

File: pkg/addr/host.go [hunk 11]
```diff
@@ -88,22 +89,28 @@ func MustParseHost(s string) Host {
...
 // Type returns the type of the address represented by h.
+// @ pure
+// @ decreases
 func (h Host) Type() HostAddrType {
 	return h.t
 }
 
 // IP returns the IP address represented by h.
 // Panics if h.Type() is not HostTypeIP.
+// @ requires h.t == HostTypeIP
```
Comment (jcp19) [inline pkg/addr/host.go]: Notice that this is a spec for a public function (thus, it needs to be understandable by clients), but it mentions a private field (which clients are not aware of). Thus, this precondition cannot be really be ever established by a client - we should change the precondition to only use public members, like the `Type()` function instead.

File: pkg/addr/host.go [hunk 12]
```diff
@@ -113,13 +120,17 @@ func (h Host) IP() netip.Addr {
 
 // SVC returns the SVC address represented by h.
 // Panics if h.Type() is not HostTypeSVC.
+// @ requires h.t == HostTypeSVC
```
Comment (jcp19) [inline pkg/addr/host.go]: same here

File: pkg/addr/host.go [hunk 13]
```diff
@@ -113,13 +120,17 @@ func (h Host) IP() netip.Addr {
...
+// @ decreases
 func (h Host) SVC() SVC {
 	if h.t != HostTypeSVC {
 		panic("SVC called on non-SVC address")
 	}
 	return h.svc
 }
 
+// @ requires (h.Type() == HostTypeNone) || (h.Type() == HostTypeIP) || (h.Type() == HostTypeSVC)
```
Comment (jcp19) [inline pkg/addr/host.go]: do we really need parentheses here?

**Comment by jcp19** [comment]
I will take a look at this PR after my comments are addressed

File: verification/dependencies/net/netip/netip.gobra (hunk 1 above)
Comment (aaronbojarski) [reply verification/dependencies/net/netip/netip.gobra]: Done.

File: verification/dependencies/net/netip/netip.gobra (hunk 2 above)
Comment (aaronbojarski) [reply verification/dependencies/net/netip/netip.gobra]: Done.

File: verification/dependencies/net/net.gobra (hunk 3 above)
Comment (aaronbojarski) [reply verification/dependencies/net/net.gobra]: Done.

File: pkg/addr/svc.go (hunk 4 above)
Comment (aaronbojarski) [reply pkg/addr/svc.go]: Done.

File: pkg/addr/isdas.go (hunk 5 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: Done.

File: pkg/addr/isdas.go (hunk 6 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: I see. Done.

File: pkg/addr/isdas.go (hunk 7 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: Done.

File: pkg/addr/isdas.go (hunk 8 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: Done.

File: pkg/addr/isdas.go (hunk 9 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: Done.

File: pkg/addr/isdas.go (hunk 10 above)
Comment (aaronbojarski) [reply pkg/addr/isdas.go]: Done.

File: pkg/addr/host.go (hunk 11 above)
Comment (aaronbojarski) [reply pkg/addr/host.go]: Done.

File: pkg/addr/host.go (hunk 12 above)
Comment (aaronbojarski) [reply pkg/addr/host.go]: Done.

File: pkg/addr/host.go (hunk 13 above)
Comment (aaronbojarski) [reply pkg/addr/host.go]: No I removed them. 
Done.

File: pkg/addr/host.go [hunk 14]
```diff
@@ -79,6 +78,8 @@ func ParseHost(s string) (Host, error) {
 
 // MustParseHost calls ParseHost(s) and panics on error.
 // It is intended for use in tests with hard-coded strings.
+// @ requires false
+// @ decreases
```
Comment (jcp19) [inline pkg/addr/host.go]:
```suggestion
```

**Review by jcp19 — APPROVED** [review]

**Comment by jcp19** [comment]
This was merged in the `verification` branch, we should merge it in the upstream PR

**Comment by aaronbojarski** [comment]
Ah I thought you wanted to keep the verification and the upstream changes in separate PRs. Thats why I created this `verification` branch. If that is not the case I will merge it into the `upstream` branch as well.

**Pull request merged**